### ISSUE TITLE
Can't stream attachment

### ISSUE THREAD

**Issue by kevinob11** [bojand/mailgun-js#102]
I'm running 0.7.11, node 6.0.0 on OS X. I can't send an attachment as a stream. Code and error are below. The stream is from a google drive file directly from the google drive node library and works great with nodemailer + mailgun SMTP.

Code
```
// This doesn't work
mailgun.messages().send({
  from: 'test@test.com',
  to: 'test@test.com',
  subject: 'test subject',
  text: 'TESTING!',
  attachment: streamFromDrive
});

// This does work
mailgun.messages().send({
  from: 'test@test.com',
  to: 'test@test.com',
  subject: 'test subject',
  text: 'TESTING!'
});
```

Error
```
Unhandled rejection Error: 'from' parameter is missing
    at IncomingMessage.<anonymous> (/Users/kevin/code/portal/api/node_modules/mailgun-js/lib/request.js:310:15)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:926:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickDomainCallback (internal/process/next_tick.js:122:9)
```

Formdata
```
FormData {
  _overheadLength: 562,
  _valueLength: 85,
  _lengthRetrievers: [],
  writable: false,
  readable: true,
  dataSize: 0,
  maxDataSize: 2097152,
  pauseStreams: true,
  _released: false,
  _streams:
   [ '----------------------------513089435233253146498067\r\nContent-Disposition: form-data; name="from"\r\n\r\n',
     'test@test.com',
     [Function: bound ],
     '----------------------------513089435233253146498067\r\nContent-Disposition: form-data; name="to"\r\n\r\n',
     'test@test.com',
     [Function: bound ],
     '----------------------------513089435233253146498067\r\nContent-Disposition: form-data; name="subject"\r\n\r\n',
     'test subject',
     [Function: bound ],
     '----------------------------513089435233253146498067\r\nContent-Disposition: form-data; name="text"\r\n\r\n',
     'TESTING!',
     [Function: bound ],
     '----------------------------513089435233253146498067\r\nContent-Disposition: form-data; name="attachment"\r\nContent-Type: application/octet-stream\r\n\r\n',
     DelayedStream {
       source: [Object],
       dataSize: 0,
       maxDataSize: Infinity,
       pauseStream: true,
       _maxDataSizeExceeded: false,
       _released: false,
       _bufferedEvents: [Object],
       _events: [Object],
       _eventsCount: 1 },
     [Function: bound ] ],
  _currentStream: null,
  _boundary: '--------------------------513089435233253146498067' }
```

**Comment by kevinob11** [comment]
Here is my stream I'm trying to attach
```
PassThrough {
  _readableState:
   ReadableState {
     objectMode: false,
     highWaterMark: 16384,
     buffer:
      [ <Buffer 25 50 44 46 2d 31 2e 34 0a 31 20 30 20 6f 62 6a 0a 3c 3c 0a 2f 54 69 74 6c 65 20 28 fe ff 29 0a 2f 43 72 65 61 74 6f 72 20 28 fe ff 29 0a 2f 50 72 6f ... >,
        <Buffer b5 90 2f 1d db be 6e fe 94 4f df 7a e6 c1 82 fe bb 11 51 84 96 6c dc 65 d8 cc b5 fb 4f 5e ba 1a 72 f4 b9 e3 3b 7f 5b 38 6d d4 e0 4e 0d cb e7 36 af 27 ... >,
        <Buffer 59 f4 c6 ec 66 e3 1f b8 7d 31 4d 6f c5 a9 a4 b7 e4 45 67 78 79 b3 5a df c2 c4 7d 62 6d 9a f2 5e c9 8d 40 78 62 06 d8 9c 83 cd 3e 42 cf 9c 64 ea 0c 26 ... >,
        <Buffer 4c 14 47 b1 5b 2f 42 7c ab 71 ec 86 11 cd 8a 05 59 5c cd 51 fb bd 65 e2 8f 12 92 6f b2 42 f4 17 56 25 cc a4 82 7a a4 bb f7 5a 2f fe 13 69 71 4b ec fb ... >,
        <Buffer f1 cb 15 09 a5 51 f3 0b 7b 88 6e c7 db 71 6f 2e ba 33 6d 26 10 f6 1a 65 26 25 88 69 45 09 24 4b f4 1b 4b fc 6a dd 0b db f1 31 82 e9 f8 04 dc 6d 3c 90 ... >,
     mailgun-js appending stream to form data. key: attachment obj: test/helpers/amazonReceipt.pdf +20ms
  mailgun-js POST /v3/praece.com/messages +1ms
     <Buffer b2 9c fb 84 5b a0 64 1a 47 95 bb 80 53 1b 6a c8 a2 99 84 88 03 a0 84 6f a4 1b 4f 8b d8 32 b7 4f c7 01 55 1a b2 81 99 19 37 e8 7e 31 7f 15 1e 40 d9 e3 ... >,
        <Buffer 11 53 05 de 35 da 4e be 71 94 8d f7 a0 0b c6 6f 00 6f 86 0a 36 de 03 1a f5 17 e9 a2 00 09 81 67 40 44 5f db 67 81 2d b6 d1 12 0a 0d e0 94 d0 41 93 6a ... >,
        <Buffer 9d c2 76 9c b5 58 bf 04 b5 65 94 6c 19 b4 71 af 27 c0 bc 41 a6 5a 1f ac 85 d2 44 53 23 6b e3 93 0d eb a4 60 91 40 7c 60 fd d8 13 92 03 67 84 12 26 86 ... >,
        <Buffer 3c 3d e9 00 52 08 8c cf 5e 79 72 04 d3 20 f5 07 6c 2b bc 09 a9 b8 1a bc 69 68 98 61 2d c4 0e a0 b6 87 25 5b f6 ba 8d 4f 15 bb 44 67 47 54 20 ed 3e 12 ... >,
        <Buffer 8c 93 6f 86 3b 28 1b 7f ce c2 40 c0 8b 03 3d e4 5b f7 a4 8d 4f 03 23 8e ce 6a 44 08 94 19 4b c1 c6 b3 c3 15 ea 7c 3d 39 fe 38 18 de 29 2f 9f 53 85 0f ... > ],
     length: 13327,
     pipes: null,
     pipesCount: 0,
     flowing: null,
     ended: false,
     endEmitted: false,
     reading: true,
     sync: false,
     needReadable: true,
     emittedReadable: true,
     readableListening: false,
     resumeScheduled: false,
     defaultEncoding: 'utf8',
     ranOut: false,
     awaitDrain: 0,
     readingMore: false,
     decoder: null,
     encoding: null },
  readable: true,
  domain: null,
  _events:
   { end: { [Function: g] listener: [Function: onend] },
     prefinish: { [Function: g] listener: [Function] },
     drain: [Function: ondrain],
     error: [Function: onerror],
     close: [Function: cleanup] },
  _eventsCount: 5,
  _maxListeners: undefined,
  _writableState:
   WritableState {
     objectMode: false,
     highWaterMark: 16384,
     needDrain: false,
     ending: false,
     ended: false,
     finished: false,
     decodeStrings: true,
     defaultEncoding: 'utf8',
     length: 0,
     writing: false,
     corked: 0,
     sync: false,
     bufferProcessing: false,
     onwrite: [Function],
     writecb: null,
     writelen: 0,
     bufferedRequest: null,
     lastBufferedRequest: null,
     pendingcb: 0,
     prefinished: false,
     errorEmitted: false,
     bufferedRequestCount: 0,
     corkedRequestsFree: CorkedRequest { next: null, entry: null, finish: [Function] } },
  writable: true,
  allowHalfOpen: true,
  _transformState:
   TransformState {
     afterTransform: [Function],
     needTransform: true,
     transforming: false,
     writecb: null,
     writechunk: null,
     writeencoding: 'buffer' } }
```

**Comment by bojand** [comment]
Hello, can you provide more info and code perhaps on how the stream `streamFromDrive ` is created. also what module / libs are used to create it. Thanks.

**Comment by kevinob11** [comment]
Sure, I sorted out what was going on here I have a work around, but I would prefer to get something a bit simpler working.

The issue is that I'm not using a normal request stream or http response stream so the form data library doesn't know the content-length of the file I'm trying to stream. If I explicitly set it using `mailgun.Attachment` it works great. The reason I'm not using a normal response or request stream is highlighted here https://github.com/request/request/issues/887 because of this issue I am streaming to a passThrough stream then passing that stream to mailgun-js.

I don't know enough about streams to know if this is even solvable, but let me know if you want more specific code given the explanation above and I can get something together tonight.

**Comment by bojand** [comment]
Hello, a request stream should work fine... something like:

```js
var request = require('request');
mailgun.messages().send({
  from: 'test@test.com',
  to: 'test@test.com',
  subject: 'test subject',
  text: 'TESTING!',
  attachment: request('http://nodejs.org/images/logo.png')
});
```

This works. Is this how you're using it?

**Comment by kevinob11** [comment]
Yeah I'm doing something a bit different. I need to catch errors prior to passing the stream to mailgun, as I frequently hit API limits in google drive and need to use exponential backoff. Example below which I based off of the recommended way of handling this from the request issue queue here: request/request#887

```javascript
const request = require('request');
const Promise = require('bluebird');
const Stream = require('stream').PassThrough;

const promise = new Promise((resolve, reject) => {
  const stream = new Stream;
  const req = request('http://nodejs.org/images/logo.png');
  req.pipe(stream);

  req.on('response', response => {
    if (response.statusCode === 200) {
      resolve(stream);
    } else if (makeSureThisIsARateLimitError) {
      resolve(doExponentialBackoff);
    } else {
      reject(response.error);
    }
  });
});

promise
  .then(stream => {
    mailgun.messages().send({
      from: 'test@test.com',
      to: 'test@test.com',
      subject: 'test subject',
      text: 'TESTING!',
      attachment: stream
    });
  });
```

**Comment by bojand** [comment]
You should be passing the `response` along as the attachment:


```js
function getLogoAndSendEmail() {
  return new Promise(function(resolve, reject) {
    const req = request('http://nodejs.org/images/logo.png');
    req.on('response', function(response) {
      if (response.statusCode === 200) {
        sendEmail(response);
      } else {
        console.log(response.statusMessage);
      }
    });
  });
}

function sendEmail(stream) {
  return mailgun.messages().send({
    from: 'test@test.com',
    to: 'me@test.com',
    subject: 'test subject',
    text: 'TESTING!',
    attachment: stream
  });
}

getLogoAndSendEmail().then(function(r) {
  console.log('all done!');
});
```

**Comment by kevinob11** [comment]
I tried passing the response and the email never finished sending. I'll give it another shot and see if that will work.

Agreed, I don't love mixing streams and promises, but I wanted to handle errors within my code for interacting with Google Drive, rather than having to handle errors individually in each place I return a stream, especially when I know I will frequently hit the rate limit and need to back off. The callback / buffer solution would work I suppose, but I figured best practice was to avoid storing the whole file in memory, but my knowledge is pretty limited in that regard, so perhaps I'm wrong about that being worth it.
